File: Java/Data Structures/Min-Stack.java
import java.util.Stack;

class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minimum;
    /** initialize your data structure here. */
    public MinStack() {
        this.stack = new Stack<>();
        this.minimum = new Stack<>();
    }

    public void push(int val) {
        if(stack.size()==0){
            stack.push(val);
            minimum.push(val);
        }else{
            if (minimum.get(minimum.size()-1)>val){
                minimum.push(val);
            } else {
                minimum.push(minimum.get(minimum.size()-1));
            }
            stack.push(val);
        }
    }

    public void pop() {
        stack.pop();
        minimum.pop();
    }

    public int top() {
        if(stack.size()==0){
            return -1;
        }
        return stack.get(stack.size()-1);
    }

    public int getMin() {
        if(minimum.size()==0){
            return -1;
        }
        return minimum.get(minimum.size()-1);
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: Python/Data Structures/Min-Stack.py
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minval = []

    def push(self, val: int) -> None:
        if len(self.stack)==0:
            self.stack.append(val)
            self.minval.append(val)
        else:
            if self.minval[-1]<val:
                self.minval.append(self.minval[-1])
            else:
                self.minval.append(val)
            self.stack.append(val)
            

    def pop(self) -> None:
        if len(self.stack)==0:
            return -1
        self.stack.pop()
        self.minval.pop()
        

    def top(self) -> int:
        if len(self.stack)==0:
            return -1
        else:
            return self.stack[-1]
            
        

    def getMin(self) -> int:
        if len(self.minval)==0:
            return -1
        else:
            return self.minval[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
